### DISCUSSION TITLE
Using "restore points"

### DISCUSSION THREAD

**Discussion by cavo789** [PostgREST/postgrest#3518] [Q&A] [answered]
Hi all

I'm thinking on a special use case, can someone tells me if it's possible or not...

I should use a BDD software (namely PHP Behat) to start functional tests on my website like, for instance, go to some page and do stuff like adding a new user then do a login with that user and f.i. create a new blog item in a given category. 

So, in short, the database will be updated (add, update or delete) when running the test. 

Once my test is validated, I wish to be able to reset my database and start another tests. This is the job of Behat...

What concern postgrest, can I "create a restore point" before each test I'll fire and once the test is done, reset the database ? 

Is there special api endpoints for this ?

I think to postgrest because before tests, I'll also need to call api for preparing the db like already create some users, blog categories, and so on.

Many, many thanks for your advices.

**Comment by wolfgangwalther** [comment, marked as answer]
This is not something that PostgREST can do. Also, I think it's pretty much out of scope for PostgREST.

That being said here are some ideas:
- If you're ok with just making a single request and have that not modify the database, look into [`db-tx-end`](https://postgrest.org/en/v12/references/transactions.html#transaction-end).
- You could also create your "restore point" as a template database in PostgreSQL and then create new databases for each test case from that template. For this to work, you'll need to be able to restart PostgREST between every test case, connecting to the new database.

**Reply by cavo789** [reply]
Thanks @wolfgangwalther for your fast answer.  No doubt it's outside the scope of PostgREST; I was just brainstorming "Can PostgREST help for this?".

I'm using with a big number of Docker containers (12 at this moment) and each application is isolated from each other so, using PostgREST is (if it was possible) an easy way to expose my database from application A / B or C; call some "Create a restore point" and allow my application D to play some CRUD statements and then "Restore" (just like if it was a transaction in fact).

Each database (A, B and C) comes with a predefined set of data (huge ones) but, between each test, I should "reset" the databases.

That's why I was brainstorming "Perhaps PostgREST already has some magic for this".

Thanks for your answer; I'll need to continue thinking on the best (and easiest) way to do.
